### PR TITLE
Add type hints to ART

### LINKED ISSUE #360: Add type hints to ART
**Describe the solution you'd like**
With the removal of support for Python 2, ART can now include type hints. These automate some of the checking logic currently present in the library and can be automatically parsed by Sphinx for docs.

### PR DESCRIPTION
**Pull request by ririnicolae**
# Description

Fixes #360

This PR adds typing to ART. The following changes were made:
- Include type hints in the entire `art` module
- Adapt docstrings to not specify variable type when this is included in the signature
- Update project requirements to include `mypy` and the modules integrating it with `sphinx`
- Include `mypy` run in Travis build
- Include `black` config file

Integration with `sphinx`: `mypy` types specified in the code are used for automatically specifying types and signatures in docs. However, when API signatures are violated in subclasses, this transforms the signature to the most generic type `func(*args, **kwargs)`. Please review generated docs and let's discuss if this is acceptable / possible solutions.

Suppressed typing errors: these can be found in the code by searching for `# type: ignore`.

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [X] Running `mypy` locally
- [X] Unit tests
- [X] Building `sphinx` docs locally

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -20,6 +20,7 @@ jobs:
         - (pycodestyle --max-line-length=120 art || exit 0)
         - (pylint --disable=C0330,C0415,E203,E1136 -rn art || exit 0)
         - py.test --pep8 -m pep8
+        - (mypy art || exit 0)
```
Comment (beat-buesser) [inline .travis.yml:23]: How about keeping the PEP8 run at the end, to make it easier to find for contributors when it fails. The other runs don't let the run fail for now.

File: art/attacks/evasion/boundary.py [hunk 2]
```diff
@@ -61,71 +62,57 @@ class BoundaryAttack(EvasionAttack):
...
         :param x_adv_init: Initial array to act as initial adversarial examples. Same shape as `x`.
         :type x_adv_init: `np.ndarray`
         :return: An array holding the adversarial examples.
-        :rtype: `np.ndarray`
         """
         y = check_and_transform_label_format(y, self.estimator.nb_classes, return_one_hot=False)
 
         # Get clip_min and clip_max from the classifier or infer them from data
-        if self.estimator.clip_values is not None:
+        if hasattr(self.estimator, "clip_values") and self.estimator.clip_values is not None:
```
Comment (beat-buesser) [inline art/attacks/evasion/boundary.py:115]: clip_values is a property of estimators.
```suggestion
        if self.estimator.clip_values is not None:
```

File: art/attacks/evasion/pixel_threshold.py [hunk 3]
```diff
@@ -152,9 +144,9 @@ def generate(self, x, y=None, maxiter=100, **kwargs):
                 self.min_th = 127
                 start, end = 1, 127
                 while True:
-                    image_result = []
+                    image_result: Union[List[np.ndarray], np.ndarray] = []
                     threshold = (start + end) // 2
-                    success, trial_image_result = self._attack(image, target_class, threshold, maxiter)
```
Comment (beat-buesser) [inline art/attacks/evasion/pixel_threshold.py:157]: Can `maxiter` really be removed?

File: art/attacks/evasion/pixel_threshold.py [hunk 4]
```diff
@@ -169,7 +161,7 @@ def generate(self, x, y=None, maxiter=100, **kwargs):
                             image_result = image
                         break
             else:
-                success, image_result = self._attack(image, target_class, self.th, maxiter)
```
Comment (beat-buesser) [inline art/attacks/evasion/pixel_threshold.py:172]: Can `maxiter` really be removed?

File: art/attacks/evasion/pixel_threshold.py [hunk 5]
```diff
@@ -195,7 +217,7 @@ def _attack_success(self, adv_x, x, target_class):
             or (not self.targeted and predicted_class != target_class)
         )
 
-    def _attack(self, image, target_class, limit, maxiter=100):
```
Comment (beat-buesser) [inline art/attacks/evasion/pixel_threshold.py:198]: Can `maxiter` really be removed?

File: art/attacks/evasion/pixel_threshold.py (hunk 3 above)
Comment (ririnicolae) [reply art/attacks/evasion/pixel_threshold.py:157]: [TODO] add it back and pass it on to optimizer in `_attack`.

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 6]
```diff
@@ -99,10 +101,16 @@ def __init__(
             preprocessing=preprocessing,
         )
 
-        assert clip_values[0] == 0, "This classifier requires un-normalized input images with clip_vales=(0, max_value)"
-        assert clip_values[1] > 0, "This classifier requires un-normalized input images with clip_vales=(0, max_value)"
-        assert preprocessing is None, "This estimator does not support `preprocessing`."
-        assert postprocessing_defences is None, "This estimator does not support `postprocessing_defences`."
+        if self.clip_values is not None:
```
Comment (beat-buesser) [inline art/estimators/object_detection/PyTorchFasterRCNN.py:104]: `clip_values` is now a property of `BaseEstimator`
```suggestion
```

File: art/estimators/object_detection/PyTorchFasterRCNN.py [hunk 7]
```diff
@@ -99,10 +101,16 @@ def __init__(
...
+            if self.clip_values[0] != 0:
+                raise ValueError("This classifier requires un-normalized input images with clip_vales=(0, max_value).")
+            if self.clip_values[1] <= 0:
+                raise ValueError("This classifier requires un-normalized input images with clip_vales=(0, max_value).")
```
Comment (beat-buesser) [inline art/estimators/object_detection/PyTorchFasterRCNN.py:108]:
```suggestion
        if self.clip_values[0] != 0:
            raise ValueError("This classifier requires un-normalized input images with clip_vales=(0, max_value).")
        if self.clip_values[1] <= 0:
            raise ValueError("This classifier requires un-normalized input images with clip_vales=(0, max_value).")
```

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+[tool.black]
+line-length=120
```
Comment (beat-buesser) [inline pyproject.toml:2]: Please add a newline at end of file.

File: tests/estimators/classification/test_keras_tf.py [hunk 9]
```diff
@@ -567,6 +567,7 @@ def test_pickle(self):
         keras_model = KerasClassifier(
             self.functional_model, clip_values=(0, 1), input_layer=1, output_layer=1, preprocessing_defences=fs
         )
+        print(keras_model.preprocessing_defences)
```
Comment (beat-buesser) [inline tests/estimators/classification/test_keras_tf.py:570]:
```suggestion
```

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 7 above)
Comment (ririnicolae) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:108]: I'm not sure this is the same, it is indeed a property, but `None` is an allowed value. In that case, trying to access one of its fields (`[0]` or `[1]`) will result in an error.

File: pyproject.toml (hunk 8 above)
Comment (ririnicolae) [reply pyproject.toml]: Done.

File: art/attacks/evasion/pixel_threshold.py (hunk 3 above)
Comment (ririnicolae) [reply art/attacks/evasion/pixel_threshold.py:157]: Added back as `max_iter`, passed to the solver.

File: art/estimators/object_detection/PyTorchFasterRCNN.py (hunk 7 above)
Comment (beat-buesser) [reply art/estimators/object_detection/PyTorchFasterRCNN.py:108]: Yes, you are right.

**Pull request merged**